### PR TITLE
Enable artifacts for s390x

### PR DESCRIPTION
**Pull request by BbolroC**
This PR enables all artifacts (container images and binaries) for `s390x` alongside `x86_64`. The self-hosted runner to build them is already registered as `s390x-runner-01` to the organisation.

The change was verified with a private runner as follows:

AS: https://github.com/BbolroC/trustee/actions/runs/8982469105
KBS: https://github.com/BbolroC/trustee/actions/runs/8982469108
KBS-client: https://github.com/BbolroC/trustee/actions/runs/8982469107

The artifacts are available at:

- https://github.com/BbolroC/trustee/pkgs/container/staged-images%2Fcoco-as-grpc
- https://github.com/BbolroC/trustee/pkgs/container/staged-images%2Fcoco-as-restful
- https://github.com/BbolroC/trustee/pkgs/container/staged-images%2Fkbs
- https://github.com/BbolroC/trustee/pkgs/container/staged-images%2Fkbs-grpc-as
- https://github.com/BbolroC/trustee/pkgs/container/staged-images%2Fkbs-client

Fixes: #364

Signed-off-by: Hyounggyu Choi <Hyounggyu.Choi@ibm.com>

### CODE REVIEW

**Comment by BbolroC** [comment]
For reviewers, this PR does not cover the full enablement for s390x. Makefiles for `AS` and `KBS` are not yet enabled because `cargo install` is used in the workflows for them. I will raise a follow-up PR to deal with this by making the affected workflows use `make`. Thanks. 😉

File: kbs/docker/Dockerfile
Comment (fitzthum) [inline kbs/docker/Dockerfile:33]: I think this is fine, but really it might make more sense to connect all this TDX-specific stuff to a TDX-related flag rather than the arch. TDX could always be set to false for s390x

File: attestation-service/attestation-service/Cargo.toml
Comment (fitzthum) [inline attestation-service/attestation-service/Cargo.toml:7]: I'm on the fence about this. Maybe it would be better for us to add a dummy s390x feature or merge @huoqifeng's PR. wdyt @Xynnn007?

File: attestation-service/attestation-service/Cargo.toml
Comment (fitzthum) [inline attestation-service/attestation-service/Cargo.toml:70]: So once the s390x verifier is in, will that only be able to be built (and run) on an s390x machine? Will we be able to support validating s390x evidence on x86?

**Review by fitzthum — COMMENTED** [review]
Looks pretty good.

You might want to check the r[elease helper script](https://github.com/confidential-containers/trustee/blob/main/hack/release-helper.sh) and make sure that it will account for the multiarch images you are addding.

File: attestation-service/attestation-service/Cargo.toml
Comment (fitzthum) [reply attestation-service/attestation-service/Cargo.toml:7]: nvm i think this is fine. hadn't read the bottom part yet.

File: .github/workflows/kbs-client-build-and-push.yaml
Comment (BbolroC) [reply .github/workflows/kbs-client-build-and-push.yaml:19]: Yep, what you are suggesting seems like:

```
jobs:
  build_and_push:
    strategy:
      matrix:
        arch:
          - x86_64
          - s390x
    env:
      RUSTC_VERSION: 1.76.0
    runs-on: ${{ matrix.arch == 'x86_64' ? 'ubuntu-22.04' : 's390x' }}
```

It looks more concise and cleaner. Thanks for the suggestion.

File: kbs/Makefile
Comment (BbolroC) [reply kbs/Makefile:5]: It sounds good. I will bring the check. Thanks!

File: attestation-service/attestation-service/Cargo.toml
Comment (BbolroC) [reply attestation-service/attestation-service/Cargo.toml:70]: Yes, once the s390x verifier comes in, if it is included in `all-verifier`, we could see x86 support validating s390x evidence, but not vice versa.

File: kbs/docker/Dockerfile
Comment (BbolroC) [reply kbs/docker/Dockerfile:33]: Yes, we could introduce a TDX variable in the build script and then set it to true/false based on the arch, even we could have more fine-grained control over different boolean variables for TEE types. But as discussed in Slack, this could be handled in a follow-up PR. wdyt, @Xynnn007 ?

Comment (Xynnn007) [reply kbs/docker/Dockerfile:33]: reasonable. In future, we need to add all plugin flags and by default set them all to true.

File: kbs/docker/Dockerfile
Comment (Xynnn007) [inline kbs/docker/Dockerfile:63]: Not related to this PR, but we need to do some optimization upon this dockerfile.

**Review by Xynnn007 — COMMENTED** [review]
Looks great. we might need to wait #345 to be merged and then merge this. because the feature flag in Cargo.toml is not prepared I think.

**Comment by huoqifeng** [comment]
> Looks great. we might need to wait #345 to be merged and then merge this. because the feature flag in Cargo.toml is not prepared I think.

@Xynnn007 might a circle dependency here, the remain work in #345 requires we can build on s390x because of the IBM SE crate. What about we merge #345 and then this PR and then add the IBM SE crate?

File: .github/workflows/as-build-and-push.yaml
Comment (BbolroC) [reply .github/workflows/as-build-and-push.yaml:38]: Good point! Thanks. Yes, we can follow your suggestion. However, when a new architecture is introduced, we'll need to revisit that condition. But if we're only checking for file existence, an ARM maintainer, for example, can simply add a script to the specified location for the self-hosted runner, eliminating the need to update their integration workflow.

As discussed in Slack, I will put a short note for why the file checking is introduced above the step name. 😉

File: attestation-service/Dockerfile.as-grpc
Comment (BbolroC) [reply attestation-service/Dockerfile.as-grpc:12]: Yeah, this could be confusing to those who are not familiar with bash. I will make it clearer with `if else fi`. Thanks!

File: kbs/docker/Dockerfile
Comment (BbolroC) [reply kbs/docker/Dockerfile:63]: Yep, we should do that as other plugins come in to build their stuff.

**Comment by BbolroC** [comment]
> > Looks great. we might need to wait #345 to be merged and then merge this. because the feature flag in Cargo.toml is not prepared I think.
> 
> @Xynnn007 might a circle dependency here, the remain work in #345 requires we can build on s390x because of the IBM SE crate. What about we merge #345 and then this PR and then add the IBM SE crate?

Thanks for the comment. I also discussed it with @Xynnn007.  We ended up with waiting until #345 is merged and updating `Cargo.toml` accordingly. 😉 Let's merge yours!!

**Comment by BbolroC** [comment]
> Looks pretty good.
> 
> You might want to check the r[elease helper script](https://github.com/confidential-containers/trustee/blob/main/hack/release-helper.sh) and make sure that it will account for the multiarch images you are addding.

It seems that container images will not be affected by this PR, but an oras artifact `kbc-client`, which is not handled by the script. So this PR does not hurt the release process. 😉

**Comment by BbolroC** [comment]
Here are the verification result for the last update:

- AS: https://github.com/BbolroC/trustee/actions/runs/8999917487
- KBS: https://github.com/BbolroC/trustee/actions/runs/8999917482
- KBS-client: https://github.com/BbolroC/trustee/actions/runs/8999917484

**Review by portersrc — APPROVED** [review]

**Review by fitzthum — APPROVED** [review]
I'm still a bit suspicious of the release script but we can fix that up afterwards if needed.

**Comment by BbolroC** [comment]
> I'm still a bit suspicious of the release script but we can fix that up afterwards if needed.

Now, I got your point. I was only considering whether the multi-arch images hurt the release helper script or not. But you wanted me to enable the script to publish the multi-arch release images. I have updated the PR. Please take a look. 😉

File: hack/release-helper.sh
Comment (portersrc) [inline hack/release-helper.sh:107]: Thanks for handling this release-helper.sh change with this PR! I wonder if there was a minor bug in the earlier script, though: It never added a `latest` tag to the new release. If you want, maybe add something like that here:
```
        docker manifest create ${ghcr_repo}/${release_pkg_name}:${latest} \
            --amend ${ghcr_repo}/${release_pkg_name}:${release_tag_full}-x86_64 \
            --amend ${ghcr_repo}/${release_pkg_name}:${release_tag_full}-s390x
        docker manifest push ${ghcr_repo}/${release_pkg_name}:${latest}
```

**Comment by BbolroC** [comment]
> Are we waiting on #345 for this?

Yep!

**Comment by genjuro214** [comment]
@BbolroC , I'm trying to build s390x kbs images based on this PR in our internal pipeline. The problem is we don't have s390x node in the pipeline. So I have to build s390x image on x86 platform.

First, I tried docker buildx with `--platform` like:
```
docker buildx build -t ${registry}/kbs:latest --platform linux/s390x --load --build-arg ARCH=s390x --build-arg HTTPS_CRYPTO=openssl . -f kbs/docker/Dockerfile
```
It works and we don't need more changes on your PR. But the performance is very low. It takes hours to build the rust in this way.

So I also tried to use rust cross-compile support, like:
```
RUSTFLAGS=" -C linker=s390x-linux-gnu-gcc" CGO_ENABLED=1 GOOS=linux GOARCH=s390x cargo install --path kbs/src/kbs --no-default-features --features coco-as-builtin,resource,openssl,opa --target "s390x-unknown-linux-gnu"
```
It also works, but requires some changes based on your PR. I think we did similar works in [guest-components](https://github.com/confidential-containers/guest-components): 
https://github.com/confidential-containers/guest-components/blob/main/attestation-agent/Makefile#L72

I will try those changes in my fork based on your fork, and raise PR after this PR is merged.

**Comment by BbolroC** [comment]
@genjuro214 thanks for your work. I think the best resolution would be not using the self-hosted runner in terms of capability and maintenance. Why don't you create a new PR based on this for the best resolution and merge it instead of this one?

**Comment by BbolroC** [comment]
@Xynnn007 Could you resume reviewing this? I have rebased this onto main as #345 is merged.

**Review by Xynnn007 — APPROVED** [review]
lgtm. Thanks!

**Pull request merged**
